File: meta-sync-scripts/common/crypto.py
"""AES-256-GCM decryption for secrets at rest.

Python port of services/meta-conversion-api/src/lib/crypto.ts. Must stay in
sync with that file's ciphertext format:

    enc:v1:<iv_b64>:<authTag_b64>:<ciphertext_b64>

Values without the "enc:v1:" prefix are legacy/dev plaintext and are
returned as-is, matching the Node implementation.
"""

import base64
from typing import Optional

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

from . import config

ENC_PREFIX = "enc:v1:"


def _get_key() -> Optional[bytes]:
    raw = config.META_ENCRYPTION_KEY
    if not raw:
        return None
    key = bytes.fromhex(raw) if len(raw) == 64 else base64.b64decode(raw)
    if len(key) != 32:
        raise RuntimeError("META_ENCRYPTION_KEY must decode to 32 bytes (64 hex chars or base64)")
    return key


def decrypt_secret(value: Optional[str]) -> str:
    if not value:
        return ""
    if not value.startswith(ENC_PREFIX):
        return value  # legacy plaintext

    key = _get_key()
    if key is None:
        raise RuntimeError("Encrypted secret present but META_ENCRYPTION_KEY is not configured")

    parts = value[len(ENC_PREFIX):].split(":")
    if len(parts) != 3:
        raise RuntimeError("Malformed encrypted secret")
    iv_b64, tag_b64, ct_b64 = parts

    iv = base64.b64decode(iv_b64)
    tag = base64.b64decode(tag_b64)
    ciphertext = base64.b64decode(ct_b64)

    # Node's crypto keeps the GCM auth tag separate from the ciphertext;
    # cryptography's AESGCM expects them concatenated (ciphertext || tag).
    aesgcm = AESGCM(key)
    plaintext = aesgcm.decrypt(iv, ciphertext + tag, None)
    return plaintext.decode("utf-8")

File: services/meta-conversion-api/src/lib/crypto.ts
import crypto from 'node:crypto';
import { config } from '../config/index.js';

// AES-256-GCM authenticated encryption for secrets at rest (Meta app_secret and
// long-lived page access_token). Ciphertext is stored as:
//   enc:v1:<iv_b64>:<authTag_b64>:<ciphertext_b64>
//
// Decryption auto-detects the prefix, so pre-existing plaintext rows keep
// working during rollout — they are returned as-is and re-encrypted on next
// write. When no key is configured, values are stored in plaintext (the key is
// required in production; see config).

const ENC_PREFIX = 'enc:v1:';

function getKey(): Buffer | null {
  const raw = config.encryptionKey;
  if (!raw) return null;
  // Accept a 64-char hex or 44-char base64 (32-byte) key.
  const key = raw.length === 64 ? Buffer.from(raw, 'hex') : Buffer.from(raw, 'base64');
  if (key.length !== 32) {
    throw new Error('META_ENCRYPTION_KEY must decode to 32 bytes (64 hex chars or base64)');
  }
  return key;
}

export function encryptSecret(plaintext: string): string {
  const key = getKey();
  if (!key) return plaintext; // no key configured — store as-is
  const iv = crypto.randomBytes(12);
  const cipher = crypto.createCipheriv('aes-256-gcm', key, iv);
  const ct = Buffer.concat([cipher.update(plaintext, 'utf-8'), cipher.final()]);
  const tag = cipher.getAuthTag();
  return `${ENC_PREFIX}${iv.toString('base64')}:${tag.toString('base64')}:${ct.toString('base64')}`;
}

export function decryptSecret(value: string | null | undefined): string {
  if (!value) return '';
  if (!value.startsWith(ENC_PREFIX)) return value; // legacy plaintext
  const key = getKey();
  if (!key) {
    throw new Error('Encrypted secret present but META_ENCRYPTION_KEY is not configured');
  }
  const [ivB64, tagB64, ctB64] = value.slice(ENC_PREFIX.length).split(':');
  if (!ivB64 || !tagB64 || !ctB64) throw new Error('Malformed encrypted secret');
  const decipher = crypto.createDecipheriv('aes-256-gcm', key, Buffer.from(ivB64, 'base64'));
  decipher.setAuthTag(Buffer.from(tagB64, 'base64'));
  return Buffer.concat([decipher.update(Buffer.from(ctB64, 'base64')), decipher.final()]).toString('utf-8');
}
